### PR TITLE
treewide: Add `UsedMemory.{try_,}virtual_address`

### LINKED ISSUE #124: Enhancement: `UsedMemory.virtual_address()`?
First of all, thanks a _ton_ for this library -- it's easily one of the most polished and user-friendly x86/AMD64 decoders I've used (and I've used quite a few).

Currently (in Rust), the `Instruction` struct has a `virtual_address` (and `try_virtual_address`) member function that takes an operand index and, if it's a memory operand, uses a callback to concretize the operand's addressing form into a specific virtual address. This is _incredibly_ handy when working at the `Instruction` level.

When working slightly below the `Instruction` level at the `InstructionInfo`/`UsedMemory` level, there's no such helper function. As such, I propose that `UsedMemory.virtual_address()` (and `try_virtual_address`) as an enhancement to that part of the API.

Here's a proposed API (just `virtual_address`, for brevity):

```rust
impl UsedMemory {
  pub fn virtual_address<F>(&self, get_register_state: F) -> u64
  where
      F: FnMut(Register) -> u64 {
      // ...
  }
}
```

...where `get_register_state` would be expected to concretize the base and index registers, provided that they're not `Register::None`. I could also throw in the VSIB parameters, where appropriate.

Let me know if this sounds interesting to you. It's something I'm happy to work on (and capable of contributing).

### PR DESCRIPTION
**Pull request by woodruffw**
Adds virtual address calculation functions, rougly comparable to the ones present on `Instruction`.

Closes #124.

### CODE REVIEW

**Comment by woodruffw** [comment]
N.B.: The API for this feature is _slightly_ different than the one on `Instruction`: it doesn't support VSIBs, and uses a generic argument to cast the effective address while performing wrapping math. For example:

```rust
// Compute a 32-bit address.
used_mem.virtual_address::<u32>(|r| whatever(r));

// Compute a 64-bit address.
used_mem.virtual_address::<u64>(|r| whatever(r));
```

I used `num_traits` (which is `no_std`) to accomplish the above. Let me know if you'd prefer another approach!

File: src/rust/iced-x86/src/info/mod.rs [hunk 1]
```diff
@@ -36,6 +36,9 @@ use super::*;
 use alloc::vec::Vec;
 use core::fmt;
 
+extern crate num_traits;
```
Comment (woodruffw) [inline src/rust/iced-x86/src/info/mod.rs:39]: N.B.: I'm not sure if this is the right place for this -- the CI complained when I put it in the top-level `lib.rs`.

I'm more familiar with Rust >= 2018 than 2015, so forgive me if this is a silly mistake 🙂

**Comment by 0xd4d** [comment]
Your code won't compile with Rust 1.20 since it doesn't support ? operator with `Option<T>` (available in Rust 1.22)

Yes I think the API should be similar to the one used by `Instruction` and should support vsib.

I also don't think it's worth adding a new crate dependency. This method should be very simple, just calculate `seg_base + ((base + index*scale + displ) & addr_mask)` where `addr_mask` is `u16::MAX`, `u32::MAX` or `u64::MAX` depending on the address size. Don't call `get_register_value` if it's `Register::None`.

This mask value can't easily be calculated now (`UsedMemory` doesn't have enough context) and will need some changes to other code, so ignore the mask for now, I'll add that code later.

**Comment by woodruffw** [comment]
> Your code won't compile with Rust 1.20 since it doesn't support ? operator with `Option<T>` (available in Rust 1.22)
> 
> Yes I think the API should be similar to the one used by `Instruction` and should support vsib.

Sounds good; I'll make those changes.

I don't think it's worth it for just this PR, but have you considered moving to a newer (>= 2018) MSRV?

> I also don't think it's worth adding a new crate dependency. This method should be very simple, just calculate `seg_base + ((base + index*scale + displ) & addr_mask)` where `addr_mask` is `u16::MAX`, `u32::MAX` or `u64::MAX` depending on the address size. Don't call `get_register_value` if it's `Register::None`.
> 
> This mask value can't easily be calculated now (`UsedMemory` doesn't have enough context) and will need some changes to other code, so ignore the mask for now, I'll add that code later.

Sounds good! That'll make this PR less useful to me until the mask gets added in, but any progress is good 😄

File: src/rust/iced-x86/src/info/mod.rs [hunk 2]
```diff
@@ -163,6 +163,87 @@ impl UsedMemory {
...
+					None => return None,
+				};
+				effective = effective.wrapping_add(base)
+			}
+		}
+
+		match self.index {
+			Register::None => {}
+			_ => {
+				let index = match get_register_value(self.index, 0, 0) {
```
Comment (woodruffw) [inline src/rust/iced-x86/src/info/mod.rs:237]: N.B.: This is where `element_index` and `element_size` go, I think, but I'm not sure how to thread those into the current `UsedMemory`. `element_index` _might_ be inferable from `memory_size`, but I'm not sure.

File: src/rust/iced-x86/src/info/mod.rs [hunk 3]
```diff
@@ -163,6 +163,87 @@ impl UsedMemory {
...
+					Some(v) => v,
+					None => return None,
+				};
+				effective = effective.wrapping_add(index << self.scale)
```
Comment (0xd4d) [inline src/rust/iced-x86/src/info/mod.rs:241]: The `scale` is a bad name here, it's actually the multiplier so you shuld multiply and not shift here.

**Review by 0xd4d — COMMENTED** [review]
> I don't think it's worth it for just this PR, but have you considered moving to a newer (>= 2018) MSRV?

Yes that will happen eventually but not this year though.

File: src/rust/iced-x86/src/info/mod.rs (hunk 3 above)
Comment (woodruffw) [reply src/rust/iced-x86/src/info/mod.rs:241]: Ah, whoops. Yeah, I had that originally and got overeager with the copy from `Instruction`. Will fix.

File: src/rust/iced-x86/src/info/mod.rs [hunk 4]
```diff
@@ -163,6 +163,87 @@ impl UsedMemory {
...
+	/// * Arg 3: `element_size`: Only used if it's a vsib memory operand. Size in bytes of elements in vector index register (4 or 8).
+	#[cfg_attr(has_must_use, must_use)]
+	#[inline]
+	pub fn try_virtual_address<F>(&self, element_index: usize, mut get_register_value: F) -> Option<u64>
+	where
+		F: FnMut(Register, usize, usize) -> Option<u64>,
+	{
+		let mut effective = self.displacement;
+
+		match self.segment {
```
[the remark this answers is not in the document]
Comment (woodruffw) [reply src/rust/iced-x86/src/info/mod.rs:212]: Will fix, thanks.

**Comment by 0xd4d** [comment]
I've added the required changes to master. You will need to pull in the latest changes and then rebase it.

Once that's done you can call `UsedMemory::address_size()` to get the address size and create the mask. If it returns `CodeSize::Unknown`, assume it's 64-bit.

`UsedMemory::vsib_size()` can be passed in as the 3rd arg to `get_register_value()`

**Comment by woodruffw** [comment]
Just made the changes. I'll test this locally as well in a bit.

File: src/rust/iced-x86/src/info/mod.rs [hunk 5]
```diff
@@ -201,6 +201,96 @@ impl UsedMemory {
...
+			_ => {
+				let index = match get_register_value(self.index, element_index, self.vsib_size.into()) {
+					Some(v) => v,
+					None => return None,
+				};
+				effective = effective.wrapping_add(index.wrapping_mul(self.scale.into()))
+			}
+		}
+
+		effective &= addr_mask;
```
Comment (0xd4d) [inline src/rust/iced-x86/src/info/mod.rs:279]: The build fails beacuse missing `use u16/u32/u64`. You can move the `match` above here instead:

```rust
		match self.address_size {
			CodeSize::Code16 => effective  = effective  as u16 as u64,
			CodeSize::Code32 => effective  = effective  as u32 as u64,
			_ => {},
		};
```

**Comment by woodruffw** [comment]
Ran these changes locally and confirmed that they work on an internal project. Please let me know if there's anything else I can do!

**Comment by 0xd4d** [comment]
Great! When the CI builds successfully I'll merge it!

**Comment by 0xd4d** [comment]
Looks like you forgot to format it, the build failed. Try running `cargo fmt`

**Comment by woodruffw** [comment]
Whoops, it's always something small 😅 

Sorry about that. Hopefully should be good to go now.

**Comment by 0xd4d** [comment]
Thanks, it's merged now!

**Comment by woodruffw** [comment]
Thanks a ton for the quick turnaround!

**Comment by woodruffw** [comment]
Sorry for the belated request here: any chance this could be part of a new release? I don't mind having a `git` dependency internally, but it'd be nice to pull it directly from crates.io 🙂

**Comment by 0xd4d** [comment]
I'll release 1.9.1, will wait until the CI is successful, see https://github.com/0xd4d/iced/pull/126

**Comment by woodruffw** [comment]
Fantastic, thanks so much!

**Comment by 0xd4d** [comment]
It's on crates.io now.

**Comment by 0xd4d** [comment]
I'll have to do it again, wrong commit.

**Comment by 0xd4d** [comment]
My bad, it's the correct commit.

**Pull request merged**